Macros/Macros.js

```diff
@@ -33,54 +35,163 @@ class FurnaceMacros {
-          const actor = token.actor;
-          const sceneId = args[1].sceneId || canvas.scene.id; 
-        const speaker = { scene: sceneId, actor: actor.id, token: args[1].tokenId, alias: token.actor.name}
-          const event = args[1].event;
-          return { game: game, ui: ui, canvas: canvas, speaker, actor, token, character: actor, sceneId, event, args: args.slice(2)}
+    static getTemplateContext(args = null, remoteContext = null) {
+        const context = {
+            game: game,
+            ui: ui,
+            canvas: canvas,
+            scene: canvas.scene,
+            args,
+            speaker: {},
+            actor: null,
+            token: null,
+            character: null
+        };
+        if (remoteContext) {
+            // Set the context based on the remote context, and make sure data is valid and the remote
+            // has a token/actor selected.
+            context.speaker = remoteContext.speaker || {};
+            if (remoteContext.actorId)
+                context.actor = game.actors.get(remoteContext.actorId) || null;
+            if (remoteContext.sceneId)
+                context.scene = game.scenes.get(remoteContext.sceneId) || canvas.scene;
+            if (remoteContext.tokenId) {
+                if (canvas.scene.id === context.scene.id) {
+                    context.token = canvas.tokens.get(remoteContext.tokenId) || null;
+                } else {
+                    const tokenData = context.scene.getEmbeddedEntity("Token", remoteContext.tokenId);
+                    if (tokenData)
+                        context.token = new Token(tokenData, context.scene);
+                }
+            }
+            if (remoteContext.characterId)
+                context.character = game.actors.get(remoteContext.characterId) || null;
+        } else {
+            context.speaker = ChatMessage.getSpeaker();
+            context.actor = game.actors.get(context.speaker.actor);
+            context.token = canvas.tokens.get(context.speaker.token);
+            context.character = game.user.character;
         }
-      
-        const speaker = ChatMessage.getSpeaker();
-        const actor = game.actors.get(speaker.actor);
-        const token = canvas.tokens.get(speaker.token);
-        const character = game.user.character;
-        return { game: game, ui: ui, canvas: canvas, speaker, actor, token, character, event, sceneId: canvas.scene.id, args }
+        return context;
     }
-    
+
+    /**
+     * Defines whether a Macro can run as a GM.
+     * For security reasons, only macros authored by the GM, and not editable by users
+     * can be run as GM
+     */
+    canRunAsGM() {
+        const author = game.users.get(this.data.author);
+        const permissions = this.data.permission || {};
+        delete permissions[this.data.author];
+        return author && author.isGM && Object.values(permissions).every(p => p < CONST.ENTITY_PERMISSIONS.OWNER)
```

Comment:
Probably should allow execution if permissions of NON GM users is < OWNER. This bit is nice by the way.

Reply:
Yeah, will do that, thanks!